src/main/java/software/amazon/ionpathextraction/SearchPathParser.java

```diff
@@ -13,88 +13,120 @@
 
 package software.amazon.ionpathextraction;
 
-import static software.amazon.ionpathextraction.utils.Preconditions.checkArgument;
+import static software.amazon.ionpathextraction.internal.Preconditions.checkArgument;
 
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.function.BiFunction;
 import software.amazon.ion.IonReader;
 import software.amazon.ion.IonType;
 import software.amazon.ion.IonWriter;
 import software.amazon.ion.system.IonReaderBuilder;
 import software.amazon.ion.system.IonTextWriterBuilder;
 import software.amazon.ionpathextraction.exceptions.PathExtractionException;
+import software.amazon.ionpathextraction.pathcomponents.AnnotatedPathComponent;
 import software.amazon.ionpathextraction.pathcomponents.Index;
 import software.amazon.ionpathextraction.pathcomponents.PathComponent;
 import software.amazon.ionpathextraction.pathcomponents.Text;
 import software.amazon.ionpathextraction.pathcomponents.Wildcard;
-import software.amazon.ionpathextraction.utils.Preconditions;
 
 /**
- * Parses a search path ion expression into {@link PathComponent}s.
+ * Parses a search path ion expression into a {@link SearchPath}s.
  */
-class PathComponentParser {
+final class SearchPathParser {
 
     private static final IonReaderBuilder READER_BUILDER = IonReaderBuilder.standard();
     private static final IonTextWriterBuilder WRITER_BUILDER = IonTextWriterBuilder.standard();
 
     private static final String WILDCARD_ESCAPE_ANNOTATION = "$ion_extractor_field";
 
     // only has static methods, should not be invoked
-    private PathComponentParser() {
+    private SearchPathParser() {
     }
 
-    static List<PathComponent> parse(final String ionPathExpression) {
-        List<PathComponent> pathComponents;
+    static <T> SearchPath<T> parse(final String ionPathExpression, final BiFunction<IonReader, T, Integer> callback) {
+        final List<PathComponent> pathComponents;
 
         try (final IonReader reader = newIonReader(ionPathExpression)) {
             checkArgument(reader.next() != null, "ionPathExpression cannot be empty");
             checkArgument(reader.getType() == IonType.SEXP || reader.getType() == IonType.LIST,
                 "ionPathExpression must be a s-expression or list");
 
+            if (reader.getTypeAnnotations().length > 0) {
+                return new AnnotatedTopLevelSearchPath<>(reader.getTypeAnnotations(), callback);
```

Comment:
Same comment about needing to support top-level annotated paths that also have components.